types.md

```diff
@@ -0,0 +1,221 @@
+output type information.
+
+## Simple Types
+Simple types can be simple primitive types, in the following table, or complex
+types described in detail after the table. All types have JSON encoding.
+
+| JSON Type Encoding | Description                                           |
+|--------------------|-------------------------------------------------------|
+| `"boolean"`        | Boolean types, in standard JSON encoding              |
+| `"date"`           | A string containing an ISO-8601 time including a zone |
+| `"floating"`       | A floating point value; precision is not specified    |
+| `"integer"`        | An integer value; precision is not specified          |
+| `"json"`           | Arbitrary JSON data that is not type checked          |
+| `"string"`         | A string                                              |
+
+- `{"is": "dictionary", "key":` _key_`, "value": ` _value_ `}`
+A dictionary/map type. This value has two possible encodings. If the _key_ type
+is a string, it _may_ be encoded as a JSON object. It can also be encoded as a
+list of lists, where the inner lists are all two elements: the key and value.
+
+- `{"is": "list", "inner":` _type_ `}`
+A list/array of items, encoded as a JSON array.
+
+- `{"is": "object", "fields": {` _f1_ `:` _t1_`,` ... `} }`
+A structure encoded as a JSON object. All fields must be present. The field
+names must follow JSON property name rules.
+
+- `{"is": "optional", "inner":` _type_ `}`
+A value which may be absent. If absent, it is encoded as `null`. Nesting this
+type has no meaningful effect; that is optional of optional of string is
+functionally identical to optional of string.
+
+- `{"is": "pair", "left":` _left_`, "right":` _right_ `}`
+A pair of values. This exists mostly as a special case of WDL. It is encoded as
+a JSON object with two properties `"left"` and `"right"`.
+
+- `{"is": "tagged-union", "options": { `_n1_ `:` _t1_`,` ... `} }`
+Create a tagged union type (or algebraic data type). The value provides one
+particular value based on the options available These are encoded as `{"type":
+`_n_`, "contents": `_v_`}` where _n_ is the name of the selected option and _v_
+is a value of the matching type.
+
+For details, see Shesmu's algebraic data types. Shesmu has the restriction that
+the types must be tuples or objects, but Vidarr does not enforce this
+restriction.
+
+- `{"is": "tuple", "elements": [` _t1_`,` _t2_`,` ... `] }`
+A heterogenous tuple of values. A tuple is effectively an object but with
+numerical indices instead of field names. It is encoded as a JSON array.
+
+## Input Types
+All of the conventions for simple types applies to input types. Additionally,
+`"file"` and `"directory"` types are supported.
+
+In both cases, the exact type will depend on the plugins on the server. The submitter can provide either:
+
+- use a file known to Vidarr by its ID: `{"type": "INTERNAL", "contents": ["`_id_`"]}`
```

Comment:
"the submitter can provide either use a file [...] or use a file" reads kind of weird

Reply:
I tried to reword this section.

Reply:
Looks good!